### ISSUE TITLE
Segmentation fault when starting a timer a second time (MacOS X backend)

### ISSUE THREAD

**Issue by astrofrog** [matplotlib/matplotlib#9655]
The following causes a segmentation fault when using the MacOS X backend:

```python
In [1]: %matplotlib osx

In [2]: import matplotlib.pyplot as plt

In [3]: fig = plt.figure()

In [4]: timer = fig.canvas.new_timer(interval=100)

In [5]: timer.single_shot = True

In [6]: def test():
   ...:     print("TEST")
   ...:     

In [7]: timer.add_callback(test)

In [8]: timer.start()

In [9]: TEST
In [9]: 

In [9]: timer.start()
Segmentation fault: 11
```

This works fine for the Qt5 backend. I'm using Matplotlib 2.1.

**Comment by dopplershift** [comment]
The problem is caused by the fact that when we set the OSX timer to single shot, it "invalidates" (i.e. frees) itself and all associated stuff. This causes us to try to decref a null pointer. Removing the decref fixes it, so "just" need to plugin a "release callback" to do the decref--let's see what we can do here...

**Comment by dopplershift** [comment]
I've got a fix up in #9662 if you want to test.

**Comment by jklymak** [comment]
Closed w #9662
